### PR TITLE
[clang-format] Add BreakBinaryOperations configuration

### LINKED ISSUE #57280: clang-format: Consider one statement per line for chained bool equalities
```c
bool f() {
  return display_id == other.display_id &&
         parent_display_id == other.parent_display_id &&
         position == other.position && offset == other.offset &&
         offset_reference == other.offset_reference;
}
```

personal preference, but it's easy to miss that the third line actually has two equalities

feel free to close if you disagree/not worth the effort

### LINKED ISSUE #58014: [clang-format] AllowBreakingBinaryOperators needed
I have been searching for this for a long time just to fine large number of people asking for this on StackOverflow and other websites, but I couldn't find a ticket on GitHub actually asking for it. 
Some related StackOverflow questions:
https://stackoverflow.com/questions/50842271/clang-format-stack-all-if-statement-arguments-if-they-are-too-long
https://stackoverflow.com/questions/48247489/prevent-clang-format-from-collapsing-multi-line-if-statements-into-a-single-line

The concept is pretty simple.
We need a Style option for the following to stay as it is. It would be ideal if it also allows it fro parameters setting and other situations where other operands are found instead of just `||` and `&&`, but the minimum would be for those two. 

```c
if( ma[index - 1] > threshold_high
 || ma[index + 1] > threshold_high
 || ma[index - cols] > threshold_high
 || ma[index + cols] > threshold_high
 || ma[index - 1 - cols] > threshold_high
 || ma[index - 1 + cols] > threshold_high
 || ma[index + 1 - cols] > threshold_high
 || ma[index + 1 + cols] > threshold_high )
```

Ideas for naming:
`AllowBreakingBinaryOperators` where it just allows this to happen without formatting the previous statement
`AlwaysBreakBinaryOperators` where every time a Binary operator is found it goes to the new line. In this case it is active formatting. (it would interact together with `BreakBeforeBinaryOperators` too choose whether to have the operator before or after the new line)

Please make this happen :)

### PR DESCRIPTION
**Pull request by ameerj**
By default, clang-format packs binary operations, but it may be desirable to have compound operations be on individual lines instead of being packed.

This PR adds the option `BreakBinaryOperations` to break up large compound binary operations to be on one line each.

This applies to all logical and arithmetic/bitwise binary operations

Maybe partially addresses #79487 ?
Closes #58014 
Closes #57280

### CODE REVIEW

File: clang/unittests/Format/FormatTest.cpp
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:27531]: Is it desired to have 5 and 6 on the same line again?

File: clang/unittests/Format/FormatTest.cpp
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:27534]: Swap the tests, first test plain LLVM style, then the modification.

File: clang/unittests/Format/FormatTest.cpp
Comment (mydeveloperday) [inline clang/unittests/Format/FormatTest.cpp:27556]: why do you need these here?

File: clang/unittests/Format/FormatTest.cpp
Comment (mydeveloperday) [inline clang/unittests/Format/FormatTest.cpp:27497]: can you test the default case

File: clang/lib/Format/ContinuationIndenter.cpp
Comment (mydeveloperday) [inline clang/lib/Format/ContinuationIndenter.cpp:1219]: which test is for this condition

File: clang/unittests/Format/FormatTest.cpp
Comment (ameerj) [reply clang/unittests/Format/FormatTest.cpp:27531]: I guess I'm indifferent to whether this happens or not, so I'll remove it from the test.

It seems to be because multiplication is a different precedence compared to addition and subtraction

File: clang/lib/Format/ContinuationIndenter.cpp
Comment (ameerj) [reply clang/lib/Format/ContinuationIndenter.cpp:1219]: This is for cases such as this:
```c++
result = operand1 +
         operand2 -
         operand3 +
         operand4 -
         operand5 +
         operand6;
```
without it, the formatting could do this:
```c++
result = operand1 +
         operand2 -
         operand3 + operand4 - operand5 + operand6;
```

File: clang/unittests/Format/FormatTest.cpp
Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp]: No, don't remove it. Fix it. ;)
As far as I understood the option tells that every operation fits in one line, or gets its own, right? Then this should apply here too.

Comment (ameerj) [reply clang/unittests/Format/FormatTest.cpp]: I'm admittedly not too familiar with the codebase. There seems to be some logic to differentiate a sequence of operations based on their precedence, and consider them to overflow the column limit or not accordingly. 

i.e. the operations inside the parens, or the multiplication do not overflow the column limit, so they don't break. But the surrounding operations that are of the addition precedence overflow the column limit, and are subject to break when BinPackBinaryOperations is false.

My attempts to fix this and have all ops on a newline are breaking inside parentheses as well, which we wouldn't want, and adding indentation where there shouldn't be. I will need some guidance if this is legitimately a deal-breaker 😅

Comment (ameerj) [reply clang/unittests/Format/FormatTest.cpp]: @HazardyKnusperkeks Would it be acceptable to specify that this would set every operation _of the same precedence_ on individual lines if they fit?
And if users report issues with the way it currently works, then we can look into resolving this

Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp]: Yeah I really think it should be done right.
It's a bit since I was in the code, but I think you need to check if you ever forced the line break for that unwrappedline, on the same parenthesis level, and then return true from mustBreak.

Comment (ameerj) [reply clang/unittests/Format/FormatTest.cpp]: Issue was related to fake parens. Latest commit resolves the bug.

**Review by HazardyKnusperkeks — COMMENTED** [review]
Could you please add a note to `clang/docs/ReleaseNotes.rst`.

**Review by HazardyKnusperkeks — APPROVED** [review]

**Comment by ameerj** [comment]
Thanks for the review and approval @HazardyKnusperkeks !

I don't have write permissions to the repo so please merge this for me when you have a moment.

File: clang/lib/Format/ContinuationIndenter.cpp
Comment (owenca) [inline clang/lib/Format/ContinuationIndenter.cpp:149]: Did you mean `Current is the right operand of a binary operator`?

File: clang/lib/Format/ContinuationIndenter.cpp
Comment (owenca) [inline clang/lib/Format/ContinuationIndenter.cpp:151]: Assert `Current.Previous` before dereferencing it.
```suggestion
  const auto &Previous = *Current.Previous;
```

File: clang/lib/Format/ContinuationIndenter.cpp
Comment (owenca) [inline clang/lib/Format/ContinuationIndenter.cpp:154]: Please remove the redundant parens.

File: clang/lib/Format/TokenAnnotator.cpp
Comment (owenca) [inline clang/lib/Format/TokenAnnotator.cpp:3158]: Remove redundant parens.

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:27650]: No trailing newline at the end of the test cases here and below.

File: clang/lib/Format/ContinuationIndenter.cpp
Comment (ameerj) [reply clang/lib/Format/ContinuationIndenter.cpp]: The wording was meant to match the comment for `startsNextParameter`, but I changed the wording to clarify a bit. 

In a case like this: `op1 * (op2 + op3)`
The function would return true for `(` since that's what starts the next operand for the `*` operation.
It would be a bit inaccurate to say `(` "`is the right operand of a binary operator`"

File: clang/lib/Format/TokenAnnotator.cpp
Comment (owenca) [reply clang/lib/Format/TokenAnnotator.cpp:3158]: The redundant parens are still present.

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [reply clang/unittests/Format/FormatTest.cpp:27650]: The trailing newlines are still present.

**Review by owenca — COMMENTED** [review]
Also, we need ConfigParseTest for the new option.

File: clang/unittests/Format/FormatTest.cpp
Comment (ameerj) [reply clang/unittests/Format/FormatTest.cpp:27650]: Sorry I misunderstood what you were asking for

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:27662]:
```suggestion
               "    operand1 + operand2 - (operand3 + operand4);",
```

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:27686]: Ditto.

Comment (ameerj) [reply clang/unittests/Format/FormatTest.cpp]: Fixed, thanks for the through review!

**Comment by owenca** [comment]
This patch doesn't seem to work with the last two examples in #79487:
```
$ cat foo.cc
template<typename T>
concept my_concept =
       std::is_object_v<T>
    && std::is_const_v<T>
    && std::default_initializable<T>;

std::uint32_t a =
      byte_buffer[0]
    | byte_buffer[1] << 8
    | byte_buffer[2] << 16
    | byte_buffer[3] << 24;
$ clang-format -style='{BinPackBinaryOperations: false, BreakBeforeBinaryOperators: NonAssignment, ColumnLimit: 50}' foo.cc
template <typename T>
concept my_concept =
    std::is_object_v<T>
    && std::is_const_v<T> && std::default_initializable<T>;

std::uint32_t a =
    byte_buffer[0]
    | byte_buffer[1] << 8 | byte_buffer[2] << 16 | byte_buffer[3] << 24;
$ 
```
The logical `&&` and bitwise `|` operations are not broken up into one per line, and the column limit is exceeded.

**Comment by owenca** [comment]
It seems that this patch implements the new option without regard to the relative precedence of the binary operators:
```
$ cat bar.cc
foooooooooooo1 = foooooooooooo2 * foooooooooooo3 +
                 foooooooooooo4 / foooooooooooo5 - foooooooooooo6;
$ clang-format -style='{BinPackBinaryOperations: false}' bar.cc
foooooooooooo1 = foooooooooooo2 *
                 foooooooooooo3 +
                 foooooooooooo4 /
                 foooooooooooo5 -
                 foooooooooooo6;
$ 
```
@mydeveloperday @HazardyKnusperkeks @rymiel are you okay with this?

**Comment by mydeveloperday** [comment]
I'm not finding the true/false every clear, can we use an enum?

**Comment by owenca** [comment]
It seems [AlignOperands](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignoperands) is supposed to do what this new option would, so maybe we should fix/extend `AlignOperands` instead?

**Comment by ameerj** [comment]
> The logical && and bitwise | operations are not broken up into one per line, and the column limit is exceeded.

@owenca this should be fixed in the latest commit

**Comment by ameerj** [comment]
> It seems [AlignOperands](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignoperands) is supposed to do what this new option would, so maybe we should fix/extend `AlignOperands` instead?

This new setting is only responsible for breaking long expressions, not aligning.
`AlignOperands` can be used in conjunction with this new setting for different alignment styles.

**Comment by ameerj** [comment]
> I'm not finding the true/false every clear, can we use an enum?

Reworked this to now be an enum. Please let me know if the intent is more clear now or if anything needs further clarification.

**Review by mydeveloperday — COMMENTED** [review]
I think the enum is clearer and leave you open to add new options as you did, thank you,

**Review by mydeveloperday — COMMENTED** [review]
I have no major objection I'll go with whatever @owenca  and @HazardyKnusperkeks  agree

**Comment by HazardyKnusperkeks** [comment]
Fine by me. The build states:

Failed Tests (1):
Clang-Unit :: Format/./FormatTests/ConfigParseTest/ParsesConfiguration

Is this from this change?

**Comment by ameerj** [comment]
> Failed Tests (1):

Fixed in the latest commit

**Comment by ameerj** [comment]
@owenca Can you review once more please?

File: clang/include/clang/Format/Format.h
Comment (owenca) [inline clang/include/clang/Format/Format.h:2241]: Should we use `BinPack` instead? I'm ok with leaving it as `Never` though.

File: clang/include/clang/Format/Format.h
Comment (owenca) [inline clang/include/clang/Format/Format.h:2262]:
```suggestion
    BBO_RespectPrecedence
```

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:27715]: Redundant.

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:27812]: Redundant.

File: clang/unittests/Format/FormatTest.cpp
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:27635]:
```suggestion
  EXPECT_EQ(Style.BreakBinaryOperations, FormatStyle::BBO_Never);

  // Logical operations
```

**Review by owenca — COMMENTED** [review]
This patch also fixes #58014 and #57280, so you can include them in the commit message.

**Review by owenca — APPROVED** [review]

**Comment by ameerj** [comment]
Thanks for the review feedback @owenca @HazardyKnusperkeks
Please merge this PR for me if you feel it is ready as I don't have merge permissions

**Comment by owenca** [comment]
> Thanks for the review feedback @owenca @HazardyKnusperkeks Please merge this PR for me if you feel it is ready as I don't have merge permissions

I wonder if anyone can come up with a better name than `OnePerLine`. See https://github.com/llvm/llvm-project/pull/101882#issuecomment-2272907960. @mydeveloperday @HazardyKnusperkeks @rymiel? If not, I can merge this patch as is.

**Comment by HazardyKnusperkeks** [comment]
Nah, I can't find a better name, which isn't a whole sentence.

**Pull request merged**
